Model/DutyCalculator.php

```diff
@@ -116,84 +116,333 @@ public function __construct(
+        $this->_logger->debug('Country is specified but state is not selected (when both are needed); so we would'.
+            'prevent call to DHL API');
+    }
+
+
+    /** Uses already (recently) retrieved duty from DHL
+     * @param bool $apply
+     * @param \Magento\Quote\Model\Quote $quote
+     * @param Magento\Quote\Api\Data\AddressInterface $address
+     */
+    public function handleCaseWhereDutyIsAlreadyRetrieved($apply, $quote, $address)
+    {
+        if ($apply) {//user indicated that (s)he wants to apply duty and tax during calculation of billing
+            $quote->setReachDuty($this->checkoutSession->getReachDuty());
+            $this->response->setDuty($this->checkoutSession->getReachDuty());
+            $this->handleTaxApplicability($address, $apply);
+            $this->_logger->debug('User wants to apply the D&T. Value of apply ::' . $apply);
+            $quote->setBaseReachDuty($this->checkoutSession->getBaseReachDuty());
+            $quote->setReachDuty($this->checkoutSession->getReachDuty());
+            $quote->setDhlQuoteId($this->checkoutSession->getDhlQuoteId());
+            $quote->setDhlBreakdown($this->checkoutSession->getDhlBreakdown());
+            $this->_logger->debug($quote->getReachDuty());
+        }
+        else {//user indicated that (s)he does not  want to apply duty and tax during calculation of billing
+            $quote->setReachDuty(0);
+            //should we reset DhlQuoteId and DhlBreakdown too so as not to break any reporting?
+            //I am assuming that we do
+            $quote->setBaseReachDuty(0);
+            $quote->setDhlQuoteId('');
+            $quote->setDhlBreakdown('');
+            $this->_logger->debug('User does not want to apply the D&T. Value of apply ::' . $apply);
+        }
+
+        $quote->save();
+        $this->response->setDuty($this->checkoutSession->getReachDuty());
+    }
+
+    /** Adjust returned duty value from DHL and additionally keep track of different data that came back from
+     * DHL API call and saves proper state of quote so that the information is available beyond this page/area
+     * @param float $duty
+     * @param Magento\Quote\Api\Data\AddressInterface $address
+     * @param bool $apply
+     * @param  Magento\Framework\App\Response $response
+     */
+    public function fillOutQuoteAndSessionUsingFeeReturned($duty, $address, $apply, $response)
+    {
+        //dealing with whether duty and tax related checkbox is selected or not
+        //or whether applying duty is a must for that country or not (a setting in magento admin
+        //panel)
+        //filling out both Quote and session state as appropriate
+        //DHL quoteID is already saved into session; so not doing it here again
+        $quote = $this->checkoutSession->getQuote();
+        $duty_adjusted = $this->priceCurrency->round($duty); //copied over pre-existing code
+        //but are we supposed to round always?
```

Comment:
you're right, it looks like this function by default rounds to zero decimal places which doesn't seem right https://www.geeksforgeeks.org/php-round-function/ I would have expected it to round to two decimal places so that it was a valid amount to charge but not to zero, that could alter the amount paid by quite a bit. This might be something we need a specific ticket for so we can see what DHL would prefer

Reply:
I was wondering about something slightly different. I was wondering about if we are always rounding then is the pricing strategy (as set on admin panel like minimum, maximum) getting applied ever? May be I have to think/look around more. Thanks.

Reply:
very good question, I thought that the pricing strategy changed the response from DHL's API and not the rounding mechanism for the value we get back from the API. But if you would like to test the different pricing strategies to see what it does that's not a bad idea

Reply:
Pretty much tried to keep previous logic intact there (how the adjustment was done to the response returned by DHL). So this exploration can be done as a separate task possibly (I feel). Hey what do you feel about that? My reasoning: how many DHL calls are made that is different issue from whether pre-existing logic/calculation was honoring pricing strategy as per spec or not. So it may be a good idea not to add more time to this ticket and focus on what it is supposed to fix (so as not to cause scope creep). Thanks! @lizzynibali

Reply:
decimals do show up, so I think I'm misreading this code. Going to close this out because I don't think we have any reason to dig into this further